From a032bc4d323a74369e7ddaff82fa1421bcf655b1 Mon Sep 17 00:00:00 2001
From: Qiang Liu <liuqiang@kylinos.cn>
Date: Tue, 23 Jun 2026 10:12:36 +0800
Subject: [PATCH 1/3] ksmbd: fix sd_ndr.data memory leak in
 ksmbd_vfs_set_sd_xattr
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

这里再描述一下ndr_encode_v4_ntacl分配内存sd_ndr.data成功，但后续操作报错，导致sd_ndr.data内存泄露。

Move kfree(sd_ndr.data) into the out cleanup path to avoid memory
leak on error.

Signed-off-by: Qiang Liu <liuqiang@kylinos.cn>
Reviewed-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/server/vfs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/smb/server/vfs.c b/fs/smb/server/vfs.c
index f5fa22d87603..8e38c748d15b 100644
--- a/fs/smb/server/vfs.c
+++ b/fs/smb/server/vfs.c
@@ -1487,8 +1487,8 @@ int ksmbd_vfs_set_sd_xattr(struct ksmbd_conn *conn,
 	if (rc < 0)
 		pr_err("Failed to store XATTR ntacl :%d\n", rc);
 
-	kfree(sd_ndr.data);
 out:
+	kfree(sd_ndr.data);
 	kfree(acl_ndr.data);
 	kfree(smb_acl);
 	kfree(def_smb_acl);
-- 
2.54.0

